File: app/src/main/java/vassar/cmpu203/vassardiningapp/view/ExpandableMealtimeAdapter.java
package vassar.cmpu203.vassardiningapp.view;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.Typeface;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.snackbar.Snackbar;

import java.util.List;

import vassar.cmpu203.vassardiningapp.databinding.MealtimeMenuBinding;
import vassar.cmpu203.vassardiningapp.model.MealtimeMenu;

public class ExpandableMealtimeAdapter extends RecyclerView.Adapter<ExpandableMealtimeAdapter.ViewHolder> {

    private List<MealtimeMenu> menus;
    private final Context context;
    private final IMenuSelectView.Listener listener;
    private final IMenuSelectView menuSelectView;
    private MealtimeMenuBinding binding;

    public ExpandableMealtimeAdapter(List<MealtimeMenu> menus, Context context, IMenuSelectView.Listener listener, IMenuSelectView menuSelectView) {
        this.menus = menus;
        this.context = context;
        this.listener = listener;
        this.menuSelectView = menuSelectView;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        binding = MealtimeMenuBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
        return new ViewHolder(binding);
    }

    @Override
    public final void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        MealtimeMenu menu = menus.get(position);
        TextView mealtimeTitle = holder.binding.mealtimeTitle;
        RecyclerView mealtimeRecycler = holder.binding.itemContainer;

        mealtimeTitle.setText(menu.getLabel());

        LinearLayoutManager layoutManager = new LinearLayoutManager(context);
        layoutManager.setOrientation(RecyclerView.VERTICAL);
        mealtimeRecycler.setLayoutManager(layoutManager);
        mealtimeRecycler.setAdapter(new MealtimeItemAdapter(menu.getMenuItems(), context, listener, menuSelectView));

        mealtimeTitle.setOnClickListener(view -> {
            mealtimeRecycler.setVisibility(
                    mealtimeRecycler.getVisibility() == View.VISIBLE ? View.GONE : View.VISIBLE
            );
            if (menu.getMenuItems().isEmpty()) {
                Snackbar.make(view, "No items to display", Snackbar.LENGTH_SHORT).show();
            }
        });
    }

    @Override
    public int getItemCount() {
        return menus.size();
    }

    @SuppressLint("NotifyDataSetChanged")
    public void setMenus(List<MealtimeMenu> menus) {
        this.menus = menus;
        notifyDataSetChanged();
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        private final MealtimeMenuBinding binding;

        public ViewHolder(@NonNull MealtimeMenuBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }
}

File: app/src/main/java/vassar/cmpu203/vassardiningapp/view/ExpandableMealtimeAdapter.kt
package vassar.cmpu203.vassardiningapp.view

import android.annotation.SuppressLint
import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.snackbar.Snackbar
import vassar.cmpu203.vassardiningapp.model.MealtimeMenu
import vassar.cmpu203.vassardiningapp.databinding.MealtimeMenuBinding

class ExpandableMealtimeAdapter(
        private var menus: List<MealtimeMenu>,
        private val context: Context,
        private val listener: IMenuSelectView.Listener,
        private val menuSelectView: IMenuSelectView
) : RecyclerView.Adapter<ExpandableMealtimeAdapter.ViewHolder>() {

    private lateinit var binding: MealtimeMenuBinding

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        binding = MealtimeMenuBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val menu = menus[position]
        val mealtimeTitle: TextView = holder.binding.mealtimeTitle
        val mealtimeRecycler: RecyclerView = holder.binding.itemContainer

        mealtimeTitle.text = menu.label

        mealtimeRecycler.layoutManager = LinearLayoutManager(context).apply {
            orientation = RecyclerView.VERTICAL
        }
        mealtimeRecycler.adapter = MealtimeItemAdapter(menu.menuItems, context, listener, menuSelectView)

        mealtimeTitle.setOnClickListener { view ->
            mealtimeRecycler.visibility = if (mealtimeRecycler.visibility == View.VISIBLE) View.GONE else View.VISIBLE
            if (menu.menuItems.isEmpty()) {
                Snackbar.make(view, "No items to display", Snackbar.LENGTH_SHORT).show()
            }
        }
    }

    override fun getItemCount(): Int = menus.size

    @SuppressLint("NotifyDataSetChanged")
    fun setMenus(menus: List<MealtimeMenu>) {
        this.menus = menus
        notifyDataSetChanged()
    }

    class ViewHolder(val binding: MealtimeMenuBinding) : RecyclerView.ViewHolder(binding.root)
}
